### PR TITLE
 rustc_mir: allow promotion of promotable temps indexed at runtime.

### LINKED ISSUE #49955: Garbage value when accessing a reference into a field/element of a const value.
All of the below code seems to be neccesary to reproduce the bug; If `ALL_THE_NUMS` is accessed directly within main, the program works as expected. 

Example:

```rust
fn get_element() -> &'static u32 {
    return &ALL_THE_NUMS[0];
}

const ALL_THE_NUMS : [u32; 1] = [
    1
];

fn main(){
    println!("Num is {}", get_element());
}
```

I expected to see this happen: `Num is 1` as the output, or a compiler error if I'm doing something wrong here (am I?)

Instead, this happened: `Num is 32766` or some other garbage value is output

## Meta

`rustc --version --verbose`:

```
rustc 1.25.0
binary: rustc
commit-hash: unknown
commit-date: unknown
host: x86_64-unknown-linux-gnu
release: 1.25.0
LLVM version: 6.0
```

Also tested on master-branch version: (same result)

```
rustc 1.27.0-dev
binary: rustc
commit-hash: unknown
commit-date: unknown
host: x86_64-unknown-linux-gnu
release: 1.27.0-dev
LLVM version: 6.0
```

### PR DESCRIPTION
**Pull request by eddyb**
Fixes #49955.

r? @nikomatsakis

### CODE REVIEW

**Comment by eddyb** [comment]
Don't r+ yet, I have at least to make the new test check itself on both MIR and the old borrowkc.

File: src/librustc_mir/transform/qualify_consts.rs [hunk 1]
```diff
@@ -748,7 +754,11 @@ impl<'a, 'tcx> Visitor<'tcx> for Qualifier<'a, 'tcx, 'tcx> {
                     if let Place::Local(local) = *place {
                         if self.mir.local_kind(local) == LocalKind::Temp {
                             if let Some(qualif) = self.temp_qualif[local] {
-                                if !qualif.intersects(Qualif::NEEDS_DROP) {
+                                // `forbidden_mut` is false, so we can safely ignore
+                                // `MUTABLE_INTERIOR` from the local's qualifications.
+                                // This allows borrowing fields which don't have
+                                // `MUTABLE_INTERIOR`, from a type that does.
```
Comment (nikomatsakis) [inline src/librustc_mir/transform/qualify_consts.rs]: It'd be nice to have an example of some Rust code where this applies, for those of us whose brains are tired

File: src/librustc_mir/transform/qualify_consts.rs [hunk 2]
```diff
@@ -722,22 +724,26 @@ impl<'a, 'tcx> Visitor<'tcx> for Qualifier<'a, 'tcx, 'tcx> {
...
                         // duplicate errors (from reborrowing, for example).
                         self.qualif = self.qualif - Qualif::MUTABLE_INTERIOR;
-                        self.add(Qualif::NOT_CONST);
                         if self.mode != Mode::Fn {
                             span_err!(self.tcx.sess, self.span, E0492,
                                       "cannot borrow a constant which may contain \
                                        interior mutability, create a static instead");
                         }
+                    } else {
+                        forbidden_mut = false;
```
Comment (nikomatsakis) [inline src/librustc_mir/transform/qualify_consts.rs]: Can you put a comment here? What does this `else` indicate? I can't see it in the diff, it'll also be hard to tell when reading. Something like:

```
// Not a mut borrow
```

(Is that right? I'm guessing)

**Review by nikomatsakis — CHANGES_REQUESTED** [review]

File: src/librustc_mir/transform/qualify_consts.rs (hunk 2 above)
Comment (eddyb) [reply src/librustc_mir/transform/qualify_consts.rs]: Hmm maybe I should split this part into its own commit, it's a straight-forward translation of the old code, but it's mixed in with new parts.

**Review by nikomatsakis — APPROVED** [review]

**Pull request merged**